docs/source-code-overview.CN.md

```diff
@@ -0,0 +1,209 @@
+- storage.h, storage.cc
+
+## Engine::Storage
+
+封装底层存储引擎（目前使用的是 `RocksDB` ）的接口, 为 Server 提供磁盘存储接口。
+涉及到的 `RocksDB` 类如下：
+- `rocksdb::DB`
+- `rocksdb::BackupEngine`
+- `rocksdb::Env`
+- `rocksdb::SstFileManager`
+- `rocksdb::RateLimiter`
+- `rocksdb::ColumnFamilyHandle`
+
+## Storage::Open
+
+1. 创建需要的 `ColumnFamily`
+2. 配置每个 `ColumnFamily`
+3. 调用 rocksdb::DB::Open() 打开 RocksDB，并统计用时
+4. 调用 `rocksdb::BackupEngine::Open()` 打开 `BackupEngine`
+
+# 服务器
+
+涉及到的文件：
+- server.h, server.cc
+- redis_cmd.cc
+- worker.h, worker.cc
+
+## Server 初始化
+
+1. 调用 Redis 命名空间的 GetCommandList 函数，并初始化命令统计
+2. 创建 Worker 和 WorkerThread, 工作线程（可以配置数目），用于处理请求
+3. 构建复制 Worker（用于主从同步），可以配置数目，也可以设置限速
+
+## Server::Start()
+1. 启动 工作线程 和 复制线程, `WorkerThread.Start()`
+2. 启动 `TaskRunner`, 用于处理异步任务 `Task`
+3. 构建并启动一个 Cron 线程
+4. 构建并启动一个 `CompactionChecker` 线程，定时手动进行 `RocksDB` 的Compaction  
+
+# 线程模型
+
+## Worker线程
+
+### Worker
+`KVRocks` 使用 `libevent` 库进行事件处理
+
+涉及到的文件:
+
+- worker.h, worker.cc
+
+工作线程 `Worker` 和 `Redis::Connection` 绑定在一起，内部用 map 存储 fd 和 Connection 的映射关系，成员函数主要是连接相关的逻辑以及event相关的对象。
+
+初始化 `Worker`（构造函数）:
+- `event_base_new()` 创建 `event_base`
+- 创建 `timer event` （默认10s检查一次）回调函数 `Worker::TimerCB()`
+- 调用 `listen()` 监听端口
+    - 监听到事件之后，将返回的 `evconnlistener` 添加到监听事件列表 `listen_events_`上，回调函数是 `Worker::newConnection()`
+
+`Worker::TimerCB()`: 定时器回调函数
+- 检查超时的 client，从 `Worker` 维护的相关数据结构中踢出
+
+`Worker::Run()` :
+- `event_base_dispatch` 启动 `event_base` 的事件循环, 处理就绪的事件
+
+`Worker::newConnection()` :
+- 获取 `bufferevent`
+- 创建 `Redis::Connection(bev, worker)`
+- 设置 `bufferevent` 的读、写、事件三种回调函数，分别为：`Redis::Connection::OnRead()`、`Redis::Connection::OnWrite()`、`Redis::Connection::OnEvent()`
+- 将 `Redis::Connection` 添加到此 `Worker` 的 `map<int, Redis::Connection*> conns_` 中
+- 设置复制 `Worker` 的限速
+
+### WorkerThread
+
+涉及到的文件:
+
+- worker.h, worker.cc
+
+`WorkerThread`: 将 thread 和 Worker 封装在一起
+
+`WorkerThread.Start()`:
+
+1. 构建并启动 thread
+2. thread 中执行 `Worker::Run()`
+
+### Redis::Connection
+
+涉及到的文件:
+- redis_connection.h, redis_connection.cc
+
+将客户端的连接抽象为`Connection`，并将一系列操作封装其中，内部使用 `libevent` 的 `eventbuffer` 做数据的读取和写入。
+
+bufferevent
+
+每个连接的socket上面会有数据，数据将存在 bufferevent 的缓冲区上，对于 bufferevent 的三种回调函数:
+- 当输入缓冲区的数据大于等于输入低水位时，读取回调就会被调用。默认情况下，输入低水位的值是 0，也就是说，只要 socket 变得可读，就会调用读取回调
+- 当输出缓冲区的数据小于等于输出低水位时，写入回调就会被调用。默认情况下，输出低水位的值是 0，也就是说，只有当输出缓冲区的数据都发送完了，才会调用写入回调。因此，默认情况下的写入回调也可以理解成为写完成了，就会调用
+- 连接关闭、连接超时或者连接发生错误时，则会调用事件回调
+
+`Connection::OnRead()`: 读取数据，查找对应命令列表，然后执行
+- 调用 Connection::Input(), 读取 bufferevent 中的内容
+- 使用 Request::Tokenize 解析，当前 Connection 中维护了 Request req_ 变量
+- 使用 Request::ExecuteCommands，执行命令
+
+`Connection::OnWrite()`: 回复客户端完毕
+- Connection::Close()，内部调用 `Worker::FreeConnection`
+
+`Connection::OnEvent()`: 处理出错、连接关闭、超时情况
+
+### Redis::Request
+
+涉及到的文件:
+- redis_request.h, redis_request.cc
+
+主要用来解析 eventbuffer 中的数据, 解析成 Redis 命令，并执行
+
+`Request::Tokenize()`：
+
+将客户端传来的数据（从 `eventbuffer` 中）读出，分隔成 Token
+
+`Request::ExecuteCommands()`：
+- 调用 `LookupCommand()` 查找命令表
```

Comment:
we have updated this implementation, moved it to `connection`.
`LookupCommand` also is updated to another one.

Reply:
Ok, I will update it right now.